"Edit In Progress" lock remains forever when a tabluar file is uploaded and going through ingest while a second file is uploaded via native API

I mentioned in standup this morning that while working on #5235 (sample data) I noticed a bug in Dataverse 4.15 having to do with the "Edit In Progress" lock that was added in pull request #5551 that I'm able to exercise using the native API.

Here's a screenshot of the lock (" Edit In Progress – Additional edits cannot be made at this time.").

![Screen Shot 2019-06-27 at 12 00 28 PM](https://user-images.githubusercontent.com/21006/60282005-f9530600-98d3-11e9-94c0-71b2f71140b4.png)

In my case, I'm trying to delete the dataset. This is only possible via API because the "Edit" button is greyed out or deactivated in the GUI. When I try to delete the dataset via the native API, I get the following error:

`{'status': 'ERROR', 'message': 'Dataset cannot be edited due to dataset lock.'}`

The dataset lock stays in place for ever. I have to manually delete it via API like this:

`curl -H "X-Dataverse-key: $ADMIN_API_TOKEN" -X DELETE "$SERVER_URL/api/datasets/$ID/locks"`

It's pretty straightforward to get the dataset into this "lock stays forever" state. Here are the steps:

- Upload a tabular file via native API. Make sure it's big enough that ingest will take a second or two.
- Upload a second file via native API. This one can be small and non-tabular. I used a tiny README.md file.

The upload of the second file will fail with the following error:

`{'status': 'ERROR', 'message': 'add.add_file_error (see logs)'}`

Since this error comes from https://github.com/IQSS/dataverse/blob/v4.15/src/main/java/edu/harvard/iq/dataverse/datasetutility/AddReplaceFileHelper.java#L1462 an EJBException was caught.

To reproduce this error I'm using code I'm working in at https://github.com/IQSS/dataverse-sample-data

I'm on this commit: https://github.com/IQSS/dataverse-sample-data/commit/ed52c31

To reproduce the bug I do the following:

- Do all the setup in https://github.com/IQSS/dataverse-sample-data/blob/ed52c316f530229b0c40463dc18c5f16d07cf11d/README.md
- Remove the `sleep(3)` at https://github.com/IQSS/dataverse-sample-data/blob/ed52c316f530229b0c40463dc18c5f16d07cf11d/create_sample_data.py#L52
- Create a second file with something like this: `echo 'This is my README.' > data/dataverses/open-source-at-harvard/datasets/open-source-at-harvard/files/README.md`
- For better consistency in reproducing the bug, replace `data/dataverses/open-source-at-harvard/datasets/open-source-at-harvard/files/2019-02-25.tsv` with a larger tabular file that will successfully ingest.
- Create the sample data with `python create_sample_data.py` as explained in the README.

Here's the most relevant part of the stack trace when the second file fails to upload:

```
[#|2019-06-27T15:59:33.740+0000|INFO|glassfish 4.1|edu.harvard.iq.dataverse.ingest.IngestServiceBean|_ThreadID=27;_ThreadName=http-listener-1(2);_TimeMillis=1561651173740;_LevelValue=800;|
  Ingest of 1 tabular data file(s) is in progress.|#]

[#|2019-06-27T15:59:33.761+0000|INFO|glassfish 4.1|edu.harvard.iq.dataverse.ingest.IngestServiceBean|_ThreadID=60;_ThreadName=p: thread-pool-1; w: 1;_TimeMillis=1561651173761;_LevelValue=800;|
  Tabular data successfully ingested; DataTable with 10 variables produced.|#]

[#|2019-06-27T15:59:33.761+0000|INFO|glassfish 4.1|edu.harvard.iq.dataverse.ingest.IngestServiceBean|_ThreadID=60;_ThreadName=p: thread-pool-1; w: 1;_TimeMillis=1561651173761;_LevelValue=800;|
  Tab-delimited file produced: /tmp/data-276393803160289888.tab|#]

[#|2019-06-27T15:59:33.860+0000|WARNING|glassfish 4.1|javax.enterprise.ejb.container|_ThreadID=28;_ThreadName=http-listener-1(3);_TimeMillis=1561651173860;_LevelValue=900;_MessageID=AS-EJB-00056;|
  A system exception occurred during an invocation on EJB EjbDataverseEngine, method: public java.lang.Object edu.harvard.iq.dataverse.EjbDataverseEngine.submit(edu.harvard.iq.dataverse.engine.command.Command) throws edu.harvard.iq.dataverse.engine.command.exception.CommandException|#]

[#|2019-06-27T15:59:33.860+0000|WARNING|glassfish 4.1|javax.enterprise.ejb.container|_ThreadID=28;_ThreadName=http-listener-1(3);_TimeMillis=1561651173860;_LevelValue=900;|
  javax.ejb.EJBException
...
	at edu.harvard.iq.dataverse.datasetutility.AddReplaceFileHelper.step_070_run_update_dataset_command(AddReplaceFileHelper.java:1450)
	at edu.harvard.iq.dataverse.datasetutility.AddReplaceFileHelper.runAddReplacePhase2(AddReplaceFileHelper.java:632)
	at edu.harvard.iq.dataverse.datasetutility.AddReplaceFileHelper.runAddReplaceFile(AddReplaceFileHelper.java:413)
	at edu.harvard.iq.dataverse.datasetutility.AddReplaceFileHelper.runAddFileByDataset(AddReplaceFileHelper.java:275)
	at edu.harvard.iq.dataverse.api.Datasets.addFileToDataset(Datasets.java:1530)
...
Caused by: javax.persistence.OptimisticLockException: Exception [EclipseLink-5010] (Eclipse Persistence Services - 2.5.2.v20140319-9ad6abd): org.eclipse.persistence.exceptions.OptimisticLockException
Exception Description: The object [edu.harvard.iq.dvn.core.study.FileMetadata[id=39]] cannot be merged because it has changed or been deleted since it was last read. 
Class> edu.harvard.iq.dataverse.FileMetadata
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.mergeInternal(EntityManagerImpl.java:544)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.merge(EntityManagerImpl.java:519)
	at com.sun.enterprise.container.common.impl.EntityManagerWrapper.merge(EntityManagerWrapper.java:305)
	at edu.harvard.iq.dataverse.engine.command.impl.UpdateDatasetVersionCommand.execute(UpdateDatasetVersionCommand.java:103)
	at edu.harvard.iq.dataverse.engine.command.impl.UpdateDatasetVersionCommand.execute(UpdateDatasetVersionCommand.java:20)
	at edu.harvard.iq.dataverse.EjbDataverseEngine.submit(EjbDataverseEngine.java:236)
...
Caused by: Exception [EclipseLink-5010] (Eclipse Persistence Services - 2.5.2.v20140319-9ad6abd): org.eclipse.persistence.exceptions.OptimisticLockException
Exception Description: The object [edu.harvard.iq.dvn.core.study.FileMetadata[id=39]] cannot be merged because it has changed or been deleted since it was last read. 
Class> edu.harvard.iq.dataverse.FileMetadata
	at org.eclipse.persistence.exceptions.OptimisticLockException.objectChangedSinceLastMerge(OptimisticLockException.java:152)
...
|#]

[#|2019-06-27T15:59:33.862+0000|SEVERE|glassfish 4.1|edu.harvard.iq.dataverse.datasetutility.AddReplaceFileHelper|_ThreadID=28;_ThreadName=http-listener-1(3);_TimeMillis=1561651173862;_LevelValue=1000;|
  add.add_file_error (see logs)|#]
```

Here is the full server.log file: [lock-bug-server.log.txt](https://github.com/IQSS/dataverse/files/3335754/lock-bug-server.log.txt)

--- a/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdateDatasetVersionCommand.java
+++ b/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdateDatasetVersionCommand.java
@@ -206,7 +206,12 @@ public class UpdateDatasetVersionCommand extends AbstractDatasetCommand<Dataset>
             }
         } finally {
             // We're done making changes - remove the lock...
+            //Failures above may occur before savedDataset is set, in which case we need to remove the lock on theDataset instead
+            if(savedDataset!=null) {
             ctxt.datasets().removeDatasetLocks(savedDataset, DatasetLock.Reason.EditInProgress);
+            } else {
+                ctxt.datasets().removeDatasetLocks(theDataset, DatasetLock.Reason.EditInProgress);
+            }
         }
         return savedDataset; 
     }

